### ISSUE TITLE
unconditional_recursion does not catch indirect infinity recursion

### ISSUE THREAD

**Issue by KisaragiEffective** [rust-lang/rust#117541]
### Code

```Rust
fn a() {
    b()
}

fn b() {
   a()
}

fn main() {
    let _ = a();
}
```

### Current output
none:
```Shell
```

### Desired output

```Shell
warning: function cannot return without recursing
 --> <source>:1:1
  |
1 | fn a() { 
  | ^^^^^^ cannot return without recursing
2 |     b()
  |     --- recursive call site
  |
  = help: a ​`loop​` may express intention better if this is on purpose
  = note: ​`#[warn(unconditional_recursion)]​` on by default

warning: function cannot return without recursing
 --> <source>:5:1
  |
5 | fn b() { 
  | ^^^^^^ cannot return without recursing
6 |     a()
  |     --- recursive call site
  |
  = help: a ​`loop​` may express intention better if this is on purpose
  = note: ​`#[warn(unconditional_recursion)]​` on by default

warning: 2 warnings emitted
```

### Rationale and extra context

The compiler will not emit warning from `unconditional_recursion` lint if the function does mutual recursion. The above case itself is obvious, but here's a bit more complicated case:

```rs
#[derive(Default)]
struct Foo {
    #[allow(dead_code)]
    b: Box<Foo>,
}

fn main() {
    let _ = Foo::default();
}
```

The derived implementation calls `Box::<Foo>::default()`, and calls `Foo::default()` again.

### Other cases
N/A

### Anything else?
1.73.0 can reproduce both case. 1.74.0-beta.4 and nightly-2023-11-02 have same issue.

**Comment by clubby789** [comment]
Duplicate of #57965 I think. It would be good to look at the previous efforts to fix this to ensure any fix doesn't result in performance regressions

**Comment by lcnr** [comment]
closing this as a duplicate. #57965 has a lot of relevant info wrt to previous approaches, it's difficult to implement this in a performant way
